Solution for task six

SvenKolaric opened the pull request.

Commit:
Solution for task six

Commit:
Changed focus from event to element WIP

Refactor eventMap to use HTMLElement as key

Commit:
Change focus from event to element FINAL

Refactored Data interface to map for simpler fetching and managing of callback func

psyburn commented on 06.event_listeners/src/eventListeners.ts:
```diff
@@ -0,0 +1,68 @@
+    element.addEventListener(event, eventHandler);
+  },
+  off(element: HTMLElement, event?: string, callback?: EventListener) {
+    if (event) {
```
you don't think it's prudent to removeEventListener after there is nothing there ?

psyburn commented on 06.event_listeners/src/eventListeners.ts:
```diff
@@ -0,0 +1,68 @@
+  },
+  off(element: HTMLElement, event?: string, callback?: EventListener) {
+    if (event) {
+      const callbackArray = elementMap.get(element)!.events.get(event);
```
will element map always have the element in question?

psyburn commented on 06.event_listeners/src/eventListeners.ts:
```diff
@@ -0,0 +1,68 @@
+    }
+    element.addEventListener(event, eventHandler);
+  },
+  off(element: HTMLElement, event?: string, callback?: EventListener) {
```
think about this function again, there are multiple things wrong here (as far as I can see) 
if you don't see things wrong here we'll talk.

psyburn commented on 06.event_listeners/src/eventListeners.ts:
```diff
@@ -0,0 +1,68 @@
+    const triggerEvent = new Event(event);
+    element.dispatchEvent(triggerEvent);
+  },
+  delegate(element: HTMLElement, className: string, eventType: string, callback: Function) {
```
any reason you're sometimes using Function and sometimes EventListener :D ? 
I WONDER what is correct .. hmmm?

psyburn commented on 06.event_listeners/src/eventListeners.ts:
```diff
@@ -0,0 +1,68 @@
+
+function eventHandler(event: Event) {
+  const { target } = event;
+  if (target instanceof HTMLElement && elementMap.has(target)) {
```
can the target be something else?

Commit:
Replace EventListeners with Functions, refactor delegate method

Commit:
Off now removes event listeners

Added removeEventsForCallbackArray method for removing events, reverted to EventListener for callbacks (and made it same for all methods)

Commit:
Add removal of event map if the callback array is empty

Commit:
Refactor the removeEvent

Commit:
Add console.warn, removed unecessary if

psyburn commented on 06.event_listeners/src/eventListeners.ts:
```diff
@@ -0,0 +1,116 @@
+              callbackData.callbacks.splice(index, 1);
+              if (callbackData.callbacks.length === 0) {
+                eventMap!.events.delete(event);
+                removeEvent(element, event, callbackData.eventHandler);
```
here you can just removeEvent(element, event) - to remove everything for that event.

Commit:
Revert changes to EventMap, removeEvent doesn't require handlerEvent any more

Commit:
Delete element if there are no events for it

Commit:
Add check if element has no events when deleting callbacks

psyburn reviewed (commented):
and now take a look at what is repeating and refactor it :)

Commit:
Add cleanupElementMap method for deleting empty events or elements from map

Commit:
Fix bug that removes other events from elements that are not intended

psyburn commented on 06.event_listeners/src/eventListeners.ts:
```diff
@@ -0,0 +1,126 @@
+          console.warn(`The given element "${element.id}" doesn't have any event of type: "${event}"`);
+        }
+      } else {
+        elementMap.forEach((elementValue, elementKey) => {
```
FYI - comment's or named private methods would have made this more soooo much easier to review

psyburn reviewed (approved):
let's go somewhere else now :D
